mgowrapper.go

```diff
@@ -0,0 +1,206 @@
+
+	info, err = tc.Upsert(selector, update)
+	return info, logAndReturnErr(sp, err)
+}
+
+func (tc tracedMgoCollection) Find(selector interface{}) MongoQuery {
+	sp, ctx := opentracing.StartSpanFromContext(tc.ctx, "find")
+	// NOTE: Find just starts the trace, the finishing call on the MongoQuery must
+	// finish it.
+	sp.LogFields(queryToFields("selector", selector)...)
+	return tracedMongoQuery{
+		q:   tc.collection.Find(selector),
+		ctx: ctx,
+	}
+}
+
+func (tc tracedMgoCollection) Remove(selector interface{}) error {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "remove")
+	sp.LogFields(queryToFields("selector", selector)...)
+	defer sp.Finish()
+
+	return logAndReturnErr(sp, tc.collection.Remove(selector))
+}
+
+func (tc tracedMgoCollection) RemoveAll(selector interface{}) (info *mgo.ChangeInfo, err error) {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "removeall")
+	sp.LogFields(queryToFields("selector", selector)...)
+	defer sp.Finish()
+
+	info, err = tc.collection.RemoveAll(selector)
+	return info, logAndReturnErr(sp, err)
+}
+
+type tracedMongoQuery struct {
+	q   *mgo.Query
+	ctx context.Context
+}
+
+func (q tracedMongoQuery) All(result interface{}) error {
+	sp := opentracing.SpanFromContext(q.ctx)
+	defer sp.Finish()
+	sp.LogFields(opentracinglog.String("access-method", "All"))
+	return logAndReturnErr(sp, q.q.All(result))
+}
+
+func (q tracedMongoQuery) One(result interface{}) (err error) {
+	sp := opentracing.SpanFromContext(q.ctx)
+	defer sp.Finish()
+	sp.LogFields(opentracinglog.String("access-method", "One"))
+	return logAndReturnErr(sp, q.q.One(result))
+}
+func (q tracedMongoQuery) Limit(n int) MongoQuery {
+	sp := opentracing.SpanFromContext(q.ctx)
+	sp.LogFields(opentracinglog.Int("query-limit", n))
+	return tracedMongoQuery{
+		q:   q.q.Limit(n),
+		ctx: q.ctx,
+	}
+}
+
+func (q tracedMongoQuery) Select(selector interface{}) MongoQuery {
+	sp := opentracing.SpanFromContext(q.ctx)
+	sp.LogFields(queryToFields("select", selector)...)
+	return tracedMongoQuery{
+		q:   q.q.Select(selector),
+		ctx: q.ctx,
+	}
+}
+
+func (q tracedMongoQuery) Hint(indexKey ...string) MongoQuery {
+	sp := opentracing.SpanFromContext(q.ctx)
+	for i, hint := range indexKey {
+		sp.LogFields(opentracinglog.String(fmt.Sprintf("hint.%d", i), hint))
+	}
+
+	return tracedMongoQuery{
+		q:   q.q.Hint(indexKey...),
+		ctx: q.ctx,
+	}
+}
+
+func (q tracedMongoQuery) Sort(fields ...string) MongoQuery {
+	sp := opentracing.SpanFromContext(q.ctx)
+	sp.LogFields(queryToFields("sort", fields)...)
+	return tracedMongoQuery{
+		q:   q.q.Sort(fields...),
+		ctx: q.ctx,
+	}
+}
+
+// logAndReturnErr is a tiny helper for adding the error to a log inline.
+func logAndReturnErr(sp opentracing.Span, err error) error {
+	sp.LogFields(opentracinglog.Error(err))
+	return err
+}
+
+// queryToFields transforms an arbitrary mgo arg into a set of log fields.
+// This is mostly geared towards bson.M, but the Sprintf fallback should handle arrays
```

Comment:
Should we truncate arrays at some point? Off the top of my head I vaguely remember AVS having some `$in: [giant array]`. I could be wrong

Reply:
I think this will be fine, in most cases we limit ourselves to sending no more than 100 ID's at once.

And more generally, this is only inputted as a log which isn't indexed by the tracing provider.... I think
It seems like they only choose to index the "tag" values rather than the "log" values.